app/src/main/java/org/mozilla/fenix/perf/ProfilerUtils.kt

```diff
@@ -0,0 +1,194 @@
+private const val PROFILER_TOKEN = "application/vnd.firefox-profiler+json;version=1.0"
+private const val TOKEN = "profileToken"
+private const val PROFILER_DATA_URL = "https://profiler.firefox.com/public/"
+
+private val firefox_features = arrayOf(
+    "screenshots", "js", "leaf", "stackwalk", "cpu", "java",
+    "processcpu", "ipcmessages", "java"
+)
+private val firefox_threads = arrayOf(
+    "GeckoMain", "Compositor", "Renderer",
+    "SwComposite", "DOM Worker"
+)
+
+private val graphics_features = arrayOf("java", "ipcmessages")
+private val graphics_threads = arrayOf(
+    "GeckoMain", "Compositor", "Renderer", "SwComposite",
+    "RenderBackend", "SceneBuilder", "WrWorker", "CanvasWorkers",
+)
+
+private val media_features = arrayOf(
+    "js", "leaf", "stackwalk", "cpu", "audiocallbacktracing",
+    "ipcmessages", "processcpu", "java"
+)
+private val media_threads = arrayOf(
+    "cubeb", "audio", "camera", "capture", "Compositor", "GeckoMain", "gmp", "graph", "grph",
+    "InotifyEventThread", "IPDL Background", "media", "ModuleProcessThread", "PacerThread",
+    "RemVidChild", "RenderBackend", "Renderer", "Socket Thread", "SwComposite",
+    "webrtc"
+)
+
+private val networking_features = arrayOf(
+    "screenshots", "js", "leaf", "stackwalk", "cpu", "java",
+    "processcpu", "ipcmessages"
+)
+private val networking_threads = arrayOf(
+    "Compositor", "DNS Resolver", "DOM Worker", "GeckoMain",
+    "Renderer", "Socket Thread", "StreamTrans", "SwComposite", "TRR Background"
+)
+
+/**
+ * Profiler settings enum for grouping features and settings together
+ */
+enum class ProfilerSettings(val threads: Array<String>, val features: Array<String>) {
+    Firefox(firefox_threads, firefox_features),
+    Graphics(graphics_threads, graphics_features),
+    Media(media_threads, media_features),
+    Networking(networking_threads, networking_features);
+}
+
+/**
+ * Collection of functionality to parse and save the profile returned by GeckoView.
+ */
+object ProfilerUtils {
+    @Throws(IOException::class)
+    private fun saveProfileUrlToClipboardAndToast(profileResult: ByteArray, context: Context): String {
+        val outputFile = createTemporaryFile(profileResult, context)
+        val response = networkCallToProfilerServer(outputFile, context)
+        val profileToken = decodeProfileToken(response)
+        outputFile.delete()
+        return PROFILER_DATA_URL + profileToken
+    }
+
+    private fun finishProfileSave(context: Context, url: String, onUrlFinish: (String) -> Unit) {
+        val clipboardManager = context.getSystemService(Context.CLIPBOARD_SERVICE) as ClipboardManager
+        val clipData = ClipData.newPlainText("Profiler URL", url)
+        clipboardManager.setPrimaryClip(clipData)
+        onUrlFinish("URL saved to clipboard successfully. $url")
+    }
+
+    private fun saveProfileAsFile(profileSaveLauncher: ActivityResultLauncher<Intent>) {
+        val intent = Intent(Intent.ACTION_CREATE_DOCUMENT).apply {
```

Comment:
question (blocking): saving a profile to a file in shared storage is also problematic because other apps on the device can read the data. Do you know how much of a privacy concern this is?

Reply:
I'm not aware. I thought of that but there was two issues. 1) If we save locally within the app files, the user can't just access it and send it through gmail / github / whatever other messaging system they use. 2) We could password save it, but then again, I'm not sure if that would be enough.